pkg/util/kubernetes/hpa/hpa.go

```diff
@@ -0,0 +1,361 @@
+				modified = append(modified, currHPA)
+			}
+			if rcvdHPA.Type == watch.Deleted {
+				deleted = append(deleted, currHPA)
+			}
+
+			watcherTimeout.Reset(c.hpaReadTimeout)
+		case <-watcherTimeout.C:
+			return new, modified, deleted, resVer, nil
+		}
+	}
+}
+
+func newHPAWatcher() *HPAWatcherClient {
+	clientAPI, err := as.GetAPIClient()
+	if err != nil {
+		log.Errorf("Error creating Client for the HPA: %s", err.Error())
+		return nil
+	}
+	hpaPollItl := config.Datadog.GetInt("hpa_watcher_polling_freq")
+	hpaRefreshItl := config.Datadog.GetInt("hpa_external_metrics_polling_freq")
+	hpaExternalMaxAge := config.Datadog.GetInt("hpa_external_metrics_max_age")
+	return &HPAWatcherClient{
+		apiClient:         clientAPI,
+		hpaReadTimeout:    100 * time.Millisecond,
+		hpaPollItl:        time.Duration(hpaPollItl) * time.Second,
+		hpaRefreshItl:     time.Duration(hpaRefreshItl) * time.Second,
+		hpaExternalMaxAge: time.Duration(hpaExternalMaxAge) * time.Second,
+	}
+}
+
+// GetHPAWatcherClient returns the HPAWatcherClient
+func GetHPAWatcherClient() *HPAWatcherClient {
+	return newHPAWatcher()
+}
+
+// HPAWatcher runs a watch process of the various HPA objects' activities to process and store the relevant info.
+// Refreshes the custom metrics stored as well.
+func (c *HPAWatcherClient) HPAWatcher() {
+	log.Info("Starting HPA Process ...")
+	tickerHPAWatchProcess := time.NewTicker(c.hpaPollItl)
+	tickerHPARefreshProcess := time.NewTicker(c.hpaRefreshItl)
+
+	var resversion string
+	err := c.createConfigMapHPA()
+	if err != nil {
+		log.Errorf("Could not create the ConfigMap %s to run the HPA process, stopping: %s", datadogHPAConfigMap, err.Error())
+		return
+	}
+
+	// Creating a leader election engine to make sure only the leader writes the metrics in the configmap and queries Datadog.
+	leaderEngine, err := leaderelection.GetLeaderEngine()
+	if err != nil {
+		log.Errorf("Could not ensure the leader election is running properly: %s", err)
+		return
+	}
+	leaderEngine.EnsureLeaderElectionRuns()
+
+	go func() {
+		for {
+			select {
+			// Ticker for the HPA Object watcher
+			case <-tickerHPAWatchProcess.C:
+				if !leaderEngine.IsLeader() {
+					continue
+				}
+				newHPA, modified, deleted, res, err := c.hpaWatcher(resversion)
+				if err != nil {
+					log.Errorf("Error while watching HPA Objects' activities: %s", err)
+					return
+				}
+				if res != resversion && res != "" {
+					resversion = res
+					if len(newHPA) > 0 {
+						c.storeHPA(newHPA)
+					}
+					if len(modified) > 0 {
+						c.storeHPA(modified)
+					}
+					if len(deleted) > 0 {
+						log.Infof("deleting if resver is diff")
+						c.removeEntryFromConfigMap(deleted)
+					}
+				}
+			// Ticker to run the refresh process for the stored external metrics
+			case <-tickerHPARefreshProcess.C:
+				if !leaderEngine.IsLeader() {
+					continue
+				}
+				c.updateCustomMetrics()
+			}
+		}
+	}()
+}
+
+func (c *HPAWatcherClient) updateCustomMetrics() error {
```

Comment:
I am okay `HPAWatcherClient` directly accessing a ConfigMap for this PR but in the future I think we could look into passing a `Store` interface to `HPAWatcherClient` instance so that we can use a more simple in memory-store for unit tests.

I imagine a very similar package to https://github.com/kubernetes/client-go/tree/master/tools/leaderelection/resourcelock that abstracts away the details of using a ConfigMap.

Reply:
Good point. Added a card in the backlog.